File: go/bob/bob.go
package bob

import (
	"strings"
	"unicode"
)

// Remark is a convenience type for identifying what kind of remark it is.
type Remark struct {
	remark string
}

func (r Remark) isSilence() bool {
	return r.remark == ""
}

func (r Remark) isShouting() bool {
	hasLetters := strings.IndexFunc(r.remark, func(c rune) bool {
		return unicode.IsLetter(c)
	}) >= 0
	isUpcased := strings.ToUpper(r.remark) == r.remark

	return hasLetters && isUpcased
}

func (r Remark) isQuestion() bool {
	return strings.HasSuffix(r.remark, "?")
}

func (r Remark) isExasperated() bool {
	return r.isShouting() && r.isQuestion()
}

func newRemark(remark string) Remark {
	return Remark{strings.TrimSpace(remark)}
}

// Hey greets with an appropriate remark.
func Hey(remark string) string {
	r := newRemark(remark)
	switch {
	case r.isSilence():
		return "Fine. Be that way!"
	case r.isExasperated():
		return "Calm down, I know what I'm doing!"
	case r.isShouting():
		return "Whoa, chill out!"
	case r.isQuestion():
		return "Sure."
	default:
		return "Whatever."
	}
}

File: clojure/bob/src/bob.clj
(ns bob,
  (:require [clojure.string :as string]))

(defn response-for [s]
  (let [remark (string/trim s)
        question? #(string/ends-with? % "?")
        forceful? #(and (re-find #"[A-z]" %) (= % (string/upper-case %)))
        exasperated? #(and (forceful? %) (question? %))
        silence? #(string/blank? %)]

    (condp apply [remark]
      silence? "Fine. Be that way!"
      exasperated? "Calm down, I know what I'm doing!"
      forceful? "Whoa, chill out!"
      question? "Sure."
      "Whatever.")))
